src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs

```diff
@@ -187,123 +187,130 @@ state switch
                 }
             }
 
-            public ICompilationTracker FreezePartialState(
-                SolutionCompilationState compilationState,
-                CancellationToken cancellationToken)
+            public ICompilationTracker FreezePartialState(CancellationToken cancellationToken)
             {
-                return FreezePartialStateWorker(
-                    compilationState, docState: null, tree: null, cancellationToken);
-            }
+                GetPartialCompilationState(
+                    documentId: null,
+                    out var state,
+                    out var inProgressProject,
+                    out var compilationPair,
+                    out var generatorInfo,
+                    cancellationToken);
 
-            public ICompilationTracker FreezePartialStateWithTree(
-                SolutionCompilationState compilationState,
-                DocumentState docState,
-                SyntaxTree tree,
-                CancellationToken cancellationToken)
-            {
-                return FreezePartialStateWorker(
-                    compilationState, docState, tree, cancellationToken);
+                // If we're already finalized then just return what we have, and with the frozen bit flipped so that
+                // any future forks keep things frozen.
+                if (state is FinalCompilationTrackerState finalState)
+                {
+                    var frozenFinalState = finalState.WithIsFrozen();
```

Comment:
>finalState [](http://example.com/codeflow?start=43&length=10)

I assume we can't just return this here too if finalstate is already frozen?

Reply:
put into: https://github.com/dotnet/roslyn/pull/72111